### PR TITLE
[CAZB-5826] Link between VCCS and Payments

### PR DESCRIPTION
**Pull request by arctindev**
## Motivation and Context

https://eaflood.atlassian.net/browse/CAZB-5826

## Checklist:

- [ ] It contains only changes required by issue (does not contain other PR)
- [ ] Includes link to an issue (if apply)
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: app/controllers/vehicles_controller.rb [hunk 1]
```diff
@@ -22,6 +23,25 @@ def enter_details
     @errors = {}
     hide_inputs
     SessionManipulation::ClearSessionDetails.call(session: session, key: 1) if vrn
+    check_params
```
Comment (likeuwill) [inline app/controllers/vehicles_controller.rb:26]: let's try to keep controller slim, try to put all logic into service, e.g.
```
redirect_to local_authority_charges_path(id: transaction_id) if service_call(params, request.referer, session)
```
Also, we need to call method in before action method, so we will not assign errors variable and etc.

File: app/services/secrets/connection.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Secrets
+  # Connection class for aws secret manager
+  class Connection < BaseService
+    def connect_to_sm
```
Comment (likeuwill) [inline app/services/secrets/connection.rb:6]: Please add description to module and class

File: app/services/secrets/vrn_encryption.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+      Base64.urlsafe_encode64(cipher.update(vrn) + cipher.final)
+    end
+
+    # Decrypting vrn
+    def decrypt_vrn(vrn)
+      decipher = cipher_decrypt
+
+      decipher.update(Base64.urlsafe_decode64(vrn)) + decipher.final
+    end
```
Comment (likeuwill) [inline app/services/secrets/vrn_encryption.rb:18]: Please keep only methods what we use, remove unnecessary methods. Also move methods to private block and keep public only what need it service call.

**Review by likeuwill — CHANGES_REQUESTED** [review]

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+##
+# Module used to redirect to select caz page
+#
+module SessionManipulation
+  # ==== Usage
+  #    SessionManipulation::ProcessToSelectCaz.call(params: params, referer: request.referer, session: session)
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:8]: No need, just describe what is going here :)

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  #
+  class ProcessToSelectCaz < BaseManipulator
+    # ==== Attributes
+    # * +session+ - the user's session
+    # * +params+ - the querry paramns
+    # * +referer+ - the last visited url
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:14]: wrong order and typos

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    #
+    def initialize(params:, referer:, session:)
+      @referer = referer
+      @session = session
+      @params = params
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:19]: u need only vrn, so u can pass to class params['vrn'] and then use as `initialize(vrn:, referer:, session:)`

File: app/services/secrets/vrn_encryption.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+# Module used to decrypt vrn from VCCS service call
+module Secrets
+  # Encryption class for vrns
+  class VrnEncryption < Connection
```
Comment (likeuwill) [inline app/services/secrets/vrn_encryption.rb:6]: `VrnDecryption` maybe?

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    end
+
+    # It checks if vrn is in request params and last visited url was from VCCS
+    # Then it adds vrn to the session
+    def call
+      return unless @params['vrn'] && @referer == "#{Rails.configuration.x.check_air_standard_url}/"
+
+      add_vrn_to_session
+      log_action('Redirecting user to: #charges/local_authority')
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:28]: no need

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      @session = session
+      @params = params
+    end
+
+    # It checks if vrn is in request params and last visited url was from VCCS
+    # Then it adds vrn to the session
+    def call
+      return unless @params['vrn'] && @referer == "#{Rails.configuration.x.check_air_standard_url}/"
+
+      add_vrn_to_session
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:27]: Try to not use `unless` operator if possible, not all language have it and it less readable to understand the logic in code
```
      return if @vrn && @referer != "#{Rails.configuration.x.check_air_standard_url}/"

      add_vrn_to_session
```

File: app/services/session_manipulation/process_to_select_caz.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+##
+# Module used to redirect to select caz page
+#
+module SessionManipulation
+  # ==== Usage
+  #    SessionManipulation::ProcessToSelectCaz.call(params: params, referer: request.referer, session: session)
+  #
+  class ProcessToSelectCaz < BaseManipulator
```
Comment (likeuwill) [inline app/services/session_manipulation/process_to_select_caz.rb:10]: `ProcessToSelectCaz` naming is most anything thing in developing, but u should try to describe what are u doing inside class. 

You're checking URL and adding to session data if the user visited the page from VCCS. So maybe  `RequestFromVccsCompliancePage`

**Review by likeuwill — CHANGES_REQUESTED** [review]

**Comment by likeuwill** [comment]
Don't forget to keep changes the same as for VCCS

File: app/controllers/vehicles_controller.rb [hunk 11]
```diff
@@ -11,6 +11,8 @@ class VehiclesController < ApplicationController # rubocop:disable Metrics/Class
   before_action :check_vrn, except: %i[enter_details submit_details not_determined]
   skip_around_action :handle_history, only: %i[enter_details submit_details]
```
Comment (likeuwill) [inline app/controllers/vehicles_controller.rb:13]: empty line

File: app/services/secrets/request_from_vccs_compliance_page.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Secrets
+  # Class used to redirect to select caz page
+  class RequestFromVccsCompliancePage < BaseService
+    # ==== Attributes
+    # * +params+ - the querry params
```
Comment (likeuwill) [inline app/services/secrets/request_from_vccs_compliance_page.rb:7]: vrn, not params

File: app/controllers/vehicles_controller.rb [hunk 13]
```diff
@@ -355,4 +356,11 @@ def hide_inputs
   def vrn_without_leading_zeros
     @vrn_without_leading_zeros ||= vrn.gsub(/^0+/, '')
   end
+
+  # Checking if vrn is in params
```
[the remark this answers is not in the document]
Comment (likeuwill) [reply app/controllers/vehicles_controller.rb:361]: still old description

File: app/services/secrets/request_from_vccs_compliance_page.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Secrets
+  # Class used to redirect to select caz page
```
Comment (likeuwill) [inline app/services/secrets/request_from_vccs_compliance_page.rb:4]: wrong description, u not doing here redirects, u just checking last request url with vccs ui and adding to session data if reuqests come from vccs

**Review by likeuwill — CHANGES_REQUESTED** [review]

**Review by dratka — APPROVED** [review]

**Review by likeuwill — APPROVED** [review]

**Pull request merged**